symengine/parser.cpp

```diff
@@ -0,0 +1,327 @@
+    {
+        vec_basic params;
+
+        while (s[iter] != ')') {
+            params.push_back(parse_string(iter+1, operator_end[iter]));
+            iter = operator_end[iter];
+        }
+
+        // make this modular, scalable
+        // make map for all those functions which take in vec_basic args, like levi_civita
+        if (params.size() == 1)
+            if (single_arg_functions.find(expr) != single_arg_functions.end())
+                return single_arg_functions[expr](params[0]);
+
+        if (params.size() == 2)
+            if (double_arg_functions.find(expr) != double_arg_functions.end())
+                return double_arg_functions[expr](params[0], params[1]);
+
+        return function_symbol(expr, params);
+
+    }
+
+    // return a <Basic> by parsing the 'expr' passed from parse_string
+    RCP<const Basic> set_result(const std::string &expr, const bool& is_not_numeric, const int& num_dots)
+    {
+        // for handling cases like "-2"
+        // expr will be "" in this case, but we must return 0
+        if (expr == "") return zero;
+
+        // if the expr wasn't numeric, it's either a constant, or a user declared symbol
+        // otherwise it's an integer (yet to add float/double support)
+        if (is_not_numeric) {
+            if (constants.find(expr) != constants.end())
+                return constants[expr];
+            return symbol(expr);
+
+        } else {
+            if (num_dots > 1) {
+                throw std::runtime_error("Invalid symbol or number!");
+
+            } else if (num_dots == 1) {
+                return real_double(std::atof(expr.c_str()));
+
+            } else {
+                return integer(std::atoi(expr.c_str()));
+            }
+        }
+    }
+
+    bool operator_error(char prev, char current)
+    {
+        if (prev == '(') {
+            if (current == ')')
+                return true;
+
+        } else if (prev == '-') {
+            if (current != '(' and current != ',' and current != ')')
+                return true;
+
+        } else {
+            if(current != ')')
+                return true;
+        }
+        return false;
+    }
+
+public:
+    // does all the preprocessing related to parsing
+    RCP<const Basic> parse(const std::string &in)
+    {
+        // stack to maintain right brackets, to match with corresponding left brackets
+        std::stack<unsigned int> right_bracket;
+        // stack to maintain operators, in order of their precedence (essentially how BODMAS was implemented)
+        std::stack<std::pair<int, unsigned int> > op_stack;
+
+        bool last_char_was_op = false;
+        char last_char = 'x';
+        s = "";
+
+        // removing spaces from the string
+        for (unsigned int i = 0; i < in.length(); ++i) {
+            if (in[i] == ' ') {
+                continue;
+
+            } else if (in[i] == '*' and in[(i+1) % in.length()] == '*') {
```

Comment:
If `in[i]` is the last character of the entered string, `in[i+1]` goes out of bounds.